### PR TITLE
modified CollectionExtensions.ToDictionary to handle nested objects

### PR DESCRIPTION
**Pull request by ali-ince**

### CODE REVIEW

File: Neo4j.Driver/Neo4j.Driver/Internal/Extensions/CollectionExtensions.cs [hunk 1]
```diff
@@ -86,8 +88,46 @@ public static string ValueToString(this object o)
...
+        {
+            foreach (var propInfo in o.GetType().GetRuntimeProperties())
+            {
+                var name = propInfo.Name;
+                var value = propInfo.GetValue(o);
+
+                if (value != null)
+                {
+                    var convertible = value as IConvertible;
+                    if (convertible != null)
```
Comment (zhenlineo) [inline Neo4j.Driver/Neo4j.Driver/Internal/Extensions/CollectionExtensions.cs]: What would happen if `covertible` is a dictionary or list?
Could you add some tests for these two situations?

Comment (ali-ince) [reply Neo4j.Driver/Neo4j.Driver/Internal/Extensions/CollectionExtensions.cs]: Yeah, missed those conditions. What about the members of arrays, lists and dictionaries? Shall we also handle conversion of their elements?

Comment (zhenlineo) [reply Neo4j.Driver/Neo4j.Driver/Internal/Extensions/CollectionExtensions.cs]: Ideally it should be a recursive call, have a look at [java driver's impl](https://github.com/neo4j/neo4j-java-driver/blob/1.5/driver/src/main/java/org/neo4j/driver/v1/Values.java#L68)

Each elements of list is a recursive call [here](https://github.com/neo4j/neo4j-java-driver/blob/1.5/driver/src/main/java/org/neo4j/driver/v1/Values.java#L195)

And according to [this](https://github.com/neo4j/neo4j-dotnet-driver/blob/1.5/Neo4j.Driver/Neo4j.Driver/Internal/IO/PackStreamWriter.cs#L40), we have to convert array to IList too

Comment (ali-ince) [reply Neo4j.Driver/Neo4j.Driver/Internal/Extensions/CollectionExtensions.cs]: Introduced recursive conversion of objects in the object hierarchy - supporting arrays, lists and dictionaries. Also added a number of integration tests that test against possible use cases documented [here](http://neo4j.com/docs/developer-manual/current/cypher/syntax/parameters/).

**Pull request merged**